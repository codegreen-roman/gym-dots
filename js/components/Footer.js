import React from 'react'

const style = {
    border: '1px solid #FF00FF',
    margin: '2rem'
}

export const Footer = () => {
    return (
        <section style={style}>Hi, I am component Footer</section>
    )
}
