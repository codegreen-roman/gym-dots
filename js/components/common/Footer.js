import React from 'react'
import { compose, merge } from 'ramda'
import { connect } from 'react-redux'
import { shape, string, func } from 'prop-types'
import { loadLocationForUsername } from '../../state/actions/creators'

const style = {
    border: '1px solid #FF00FF',
    margin: '2rem'
}

const blockStyle = status => {

    const mergeStyle = merge(style)

    return mergeStyle({
        opacity: status === 'ready' ? '0.9' : '0.1'
    })
}

const _Footer = ({ githubLocation, fireAction }) => {

    const handleClick = () => fireAction(githubLocation.username)

    return (
        <section style={blockStyle(githubLocation.status)}>
            Hi, I am component Footer
            <button onClick={handleClick}>Request</button>
            <div>{githubLocation.username}</div>
            <div>{githubLocation.location}</div>
        </section>
    )
}

_Footer.propTypes = {
    githubLocation: shape({
        username: string.isRequired,
        location: string.isRequired,
        status: string
    }),
    fireAction: func.isRequired
}

const mapStateToProps = ({ githubLocation }) => ({
    githubLocation
})

const mapActionsToProps = dispatch => {
    return {
        fireAction: compose(dispatch, loadLocationForUsername)
    }
}

export { _Footer }
export const Footer = connect(mapStateToProps, mapActionsToProps)(_Footer)
