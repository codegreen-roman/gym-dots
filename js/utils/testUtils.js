export const findByDataAttr = (wrapper, value) => {
    return wrapper.find(`[data-test='${value}']`)
}
