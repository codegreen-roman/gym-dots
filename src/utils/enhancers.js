import React from 'react'
import { compose, branch, renderComponent, withState, withHandlers } from 'recompose'
import R from 'ramda'


/**
 * When matches predicate renderComponent
 * Third argument of branch is default to the identity function
 */
export const renderWhen = ({ when, render }) => {
    return branch(when, renderComponent(render))
}

/**
 * Render non-optimal states
 */
export const nonOptimalStates = states => {
    return compose(...states.map(renderWhen))
}

/**
 * Auxiliary element. A self-eradicating component for rendering multiple elements.
 * Example: const Root = () => return <Aux><div>1</div><div>2</div></Aux>
 * Benefit: wrapping <div> is not needed, DOM much cleaner
 */
export const Aux = props => props.children

/**
 * Add toggle functionality via HOC
 */
export const withToggle = compose(
    withState('toggledOn', 'toggle', false),
    withHandlers({
        show: ({ toggle }) => () => toggle(true),
        hide: ({ toggle }) => () => toggle(false),
        toggle: ({ toggle }) => () => toggle((current) => !current)
    })
)

/* Composition helpers */
/* eslint-disable react/display-name */
const combine = R.curry((c, o) => x => (
    <div>{c(x)} {o(x)}</div>)
)

export const combineComponents = (...args) => {
    const [first, ...rest] = args
    return R.reduce((acc, c) => combine(acc, c), first, rest)
}
