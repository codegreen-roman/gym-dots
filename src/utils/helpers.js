import {
    curry,
    compose,
    not,
    isEmpty,
    addIndex,
    map,
    prop
} from 'ramda'
/**
 * branch
 *
 * this function accepts a predicate function
 * and returns the first component if true, and the second
 * if predicate is false
 *
 * branch :: (Function -> Bool) -> Element -> Element -> Elment
 *
 * @param {function|boolean} [predicatefn] function that returns a boolean or a boolean
 * @param {node}             [comp1]       element to render if predicate is true
 * @param {node}             [comp2]       element to render if predicate is false
 * @returns {node}                         returns the correct component according to predicate
 */
export const branch = curry((predicatefn, comp1, comp2) => (predicatefn ? comp1 : comp2)); // eslint-disable-line

/**
 * Function (Component) that returns null.
 * Example usage: branch(isEmpty(items), <RenderNothing />, <Item items={items} />)
 */
export const RenderNothing = () => null

/**
 * Opposite of isEmpty
 */
export const notEmpty = compose(not, isEmpty)
// tried const notEmpty = complement(isEmpty), which is the same, but test ManageExerciseList.helper.spec.js:8 failed =(

/**
 * Adding indexes to map
 */
export const mapIndexed = addIndex(map)

/**
 * Exctracting event.target.value
 */
export const extractTargetValue = compose(prop('value'), prop('target'))
