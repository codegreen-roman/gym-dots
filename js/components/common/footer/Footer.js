import React from 'react'
import { connect } from 'react-redux'
import { FooterButton } from './buttons/FooterButton'
import { footerStyle } from './Footer.glamor'
import { startWorkoutWithCountdown, loadAppDefaults, subscribeToAppDefaultsChanges } from '../../../state/actions/index'
import { compose } from 'ramda'
import { func, bool } from 'prop-types'
import { Flex } from 'glamorous-jsxstyle'

export class _Footer extends React.Component {

    componentDidMount() {
        const { loadDefaults } = this.props
        loadDefaults()
    }

    renderTrainingButtons() {

        const { fireStartWorkout, blocked, training } = this.props
        const buttonTitle = blocked ? 'Starting ...' : 'Start Workout'

        if (training) {
            return (
                <Flex>
                    <FooterButton data-test='failButton' disabled={false}>Failed</FooterButton>
                    <FooterButton data-test='doneButton' disabled={false}>Done</FooterButton>
                </Flex>
            )
        }

        return (
            <Flex>
                <FooterButton disabled={blocked} clickHandler={fireStartWorkout}>{buttonTitle}</FooterButton>
            </Flex>
        )
    }

    render() {

        const { hidden } = this.props

        if (hidden) return null

        return (
            <footer {...footerStyle}>
                {this.renderTrainingButtons()}
            </footer>
        )
    }
}

_Footer.propTypes = {
    blocked: bool.isRequired,
    training: bool.isRequired,
    hidden: bool.isRequired,
    fireStartWorkout: func.isRequired,
    loadDefaults: func.isRequired
}

const mapStateToProps = ({ workoutStatus }) => ({
    hidden: false,
    blocked: workoutStatus === 'starting',
    training: workoutStatus === 'started',
})

const mapActionsToProps = (dispatch, { history }) => {

    const startWorkout = compose(dispatch, startWorkoutWithCountdown)

    subscribeToAppDefaultsChanges(dispatch)

    return {
        fireStartWorkout: () => {
            startWorkout()
                .then(() => history.push('/activity/workout'))
        },
        loadDefaults: compose(dispatch, loadAppDefaults)
    }
}

export const Footer = connect(mapStateToProps, mapActionsToProps)(_Footer)
