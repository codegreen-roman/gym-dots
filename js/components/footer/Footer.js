import React from 'react'
import { connect } from 'react-redux'
import glamorous from 'glamorous'
import { footerStyle } from './Footer.glamor'
import { startWorkoutWithCountdown, loadAppDefaults, subscribeToAppDefaultsChanges } from '../../state/actions'
import { compose } from 'ramda'
import { func, bool, object } from 'prop-types'

const AButton = glamorous.button({
    flex: 1
})

export class _Footer extends React.Component {

    componentDidMount() {
        const { loadDefaults } = this.props
        loadDefaults()
    }

    render() {

        const { fireStartWorkout, blocked, history } = this.props
        const handleClick = () => {
            fireStartWorkout()
                .then(() => history.push('/workout/started'))
        }

        const buttonTitle = blocked ? 'Starting ...' : 'Start Workout'

        return (
            <footer {...footerStyle}>
                <AButton disabled={blocked} onClick={handleClick}>{buttonTitle}</AButton>
            </footer>
        )
    }
}

_Footer.propTypes = {
    history: object,
    blocked: bool.isRequired,
    fireStartWorkout: func.isRequired,
    loadDefaults: func.isRequired
}

const mapStateToProps = ({ workoutStatus }) => ({
    blocked: workoutStatus === 'starting'
})

const mapActionsToProps = dispatch => {

    subscribeToAppDefaultsChanges(dispatch)

    return {
        fireStartWorkout: compose(dispatch, startWorkoutWithCountdown),
        loadDefaults: compose(dispatch, loadAppDefaults)
    }
}

export const Footer = connect(mapStateToProps, mapActionsToProps)(_Footer)
