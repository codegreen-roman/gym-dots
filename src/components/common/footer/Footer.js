import React from 'react'
import { FooterButton } from './buttons/FooterButton'
import { footerStyle } from './Footer.glamor'
import { func, bool, object, array } from 'prop-types'
import { Flex } from 'glamorous-jsxstyle'

export class Footer extends React.Component {

    componentWillReceiveProps({ shouldEndExercise, fireCompleteExercise, nextExercise: { exerciseKey }, currentResults }) {
        if (shouldEndExercise) {
            return fireCompleteExercise(exerciseKey, currentResults)
        }
    }

    renderTrainingButtons() {

        const { fireStartWorkout, blocked, training, onSetFailed, onSetDone, nextExercise } = this.props
        const buttonTitle = blocked ? 'Starting ...' : 'Start Workout'

        const startButtonHandler = () => {
            fireStartWorkout(nextExercise)
        }

        if (training) {
            return (
                <Flex>
                    <FooterButton clickHandler={onSetFailed} data-test='failButton' disabled={false}>Failed</FooterButton>
                    <FooterButton clickHandler={onSetDone} data-test='doneButton' disabled={false}>Done</FooterButton>
                </Flex>
            )
        }

        return (
            <Flex>
                <FooterButton disabled={blocked} clickHandler={startButtonHandler}>{buttonTitle}</FooterButton>
            </Flex>
        )
    }

    render() {

        const { hidden } = this.props

        if (hidden) return null

        return (
            <footer {...footerStyle}>
                {this.renderTrainingButtons()}
            </footer>
        )
    }
}

Footer.propTypes = {
    blocked: bool.isRequired,
    training: bool.isRequired,
    hidden: bool.isRequired,
    fireStartWorkout: func.isRequired,
    fireCompleteExercise: func.isRequired,
    onSetFailed: func.isRequired,
    onSetDone: func.isRequired,
    shouldEndExercise: bool.isRequired,
    nextExercise: object,
    currentResults: array.isRequired
}
