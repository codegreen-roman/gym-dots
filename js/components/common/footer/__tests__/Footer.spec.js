import React from 'react'
import { _Footer as Footer } from '../Footer'
import { mount } from 'enzyme'
import { findByDataAttr } from '../../../../utils/testUtils'

const setup = (blocked = false, training = false) => {

    const props = {
        fireStartWorkout: jest.fn(),
        loadDefaults: jest.fn(),
        blocked,
        training
    }

    return {
        component: mount(<Footer {...props} />),
        props
    }
}

describe('Footer component', () => {

    describe('snapshot not blocked', () => {
        it('matches the previous snapshot', () => {
            const { component } = setup()
            expect(component).toMatchSnapshot()
        })
    })

    describe('snapshot blocked', () => {
        it('matches the previous snapshot', () => {
            const { component } = setup(true)
            expect(component).toMatchSnapshot()
        })
    })

    describe('clicking on AButton', () => {
        it('should call fireStartWorkout function', () => {
            const { component, props } = setup()
            component.find('button').simulate('click')

            expect(props.fireStartWorkout).toHaveBeenCalledTimes(1)
        })
    })

    describe('.componentDidMount', () => {
        it('should call the loadDefaults', () => {
            const { props } = setup()
            expect(props.loadDefaults).toHaveBeenCalledTimes(1)
        })
    })

    describe('in training status', () => {

        it('should show 2 buttons', () => {
            const { component } = setup(false, true)
            const buttons = component.find('button')

            expect(buttons.length).toBe(2)

        })

        it('show failed button', () => {
            const { component } = setup(false, true)
            const button = findByDataAttr(component, 'failButton')

            expect(button.text()).toBe('Failed')

        })

        it('show done button', () => {
            const { component } = setup(false, true)
            const button = findByDataAttr(component, 'doneButton')

            expect(button.text()).toBe('Done')

        })

        it('matches the previous snapshot', () => {
            const { component } = setup(false, true)
            expect(component).toMatchSnapshot()
        })
    })

})
