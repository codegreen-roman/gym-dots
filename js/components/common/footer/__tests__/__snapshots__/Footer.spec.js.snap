// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Footer component in training status matches the previous snapshot 1`] = `
<_Footer
  blocked={false}
  fireStartWorkout={[Function]}
  loadDefaults={[Function]}
  training={true}
>
  <footer
    data-css-116hn21=""
  >
    <glamorous(div)>
      <div
        className="css-1cqgl9p"
      >
        <glamorous(button)
          data-test="failButton"
          disabled={false}
        >
          <button
            className="css-mfy564"
            data-test="failButton"
            disabled={false}
          >
            Failed
          </button>
        </glamorous(button)>
        <glamorous(button)
          data-test="doneButton"
          disabled={false}
        >
          <button
            className="css-mfy564"
            data-test="doneButton"
            disabled={false}
          >
            Done
          </button>
        </glamorous(button)>
      </div>
    </glamorous(div)>
  </footer>
</_Footer>
`;

exports[`Footer component snapshot blocked matches the previous snapshot 1`] = `
<_Footer
  blocked={true}
  fireStartWorkout={[Function]}
  loadDefaults={[Function]}
  training={false}
>
  <footer
    data-css-116hn21=""
  >
    <glamorous(button)
      disabled={true}
      onClick={[Function]}
    >
      <button
        className="css-mfy564"
        disabled={true}
        onClick={[Function]}
      >
        Starting ...
      </button>
    </glamorous(button)>
  </footer>
</_Footer>
`;

exports[`Footer component snapshot not blocked matches the previous snapshot 1`] = `
<_Footer
  blocked={false}
  fireStartWorkout={[Function]}
  loadDefaults={[Function]}
  training={false}
>
  <footer
    data-css-116hn21=""
  >
    <glamorous(button)
      disabled={false}
      onClick={[Function]}
    >
      <button
        className="css-mfy564"
        disabled={false}
        onClick={[Function]}
      >
        Start Workout
      </button>
    </glamorous(button)>
  </footer>
</_Footer>
`;
